### DISCUSSION TITLE
Clarification questions about versionScheme and eviction errors

### DISCUSSION THREAD

**Discussion by sjrd** [sbt/sbt#6572] [Q&A] [answered]
We are in the process of setting up `versionScheme` in the Scala.js repo (see https://github.com/scala-js/scala-js/pull/4517). That repo publishes a number of artifacts, with different compatibility guarantees and complex interactions. In order to decide what `versionScheme` to use for what artifact, I would like to ask some clarification questions about how `versionScheme` and eviction errors truly work.

## What if `versionScheme` is different in v1.3 and v1.5 of a lib A?

A v1.3 defines `versionScheme := Some("x")` but v1.5 defines `versionScheme := Some("y")`.

If A v1.3 is evicted in favor of A v1.5, **what version scheme is used to handle eviction errors**? I am hoping that `Some("y")` is used.

*Use case:* the "severe" breakages of our version policy, for which we might need to decide in the future whether they become major or minor bump.

## What about "fake" evictions?

Consider the following setup:
```
┌────────┐        ┌────────┐
│ A v1.3 │        │ A v1.5 │
└────▲───┘        └───▲────┘
     │                │
     │                │
┌────┴───┐        ┌───┴────┐
│ B v1.3 │        │ B v1.5 │
└────▲───┘        └───▲────┘
     │                │
     │                │    ┌────────┐
     │                └────┤ X v2.1 │
     │                     └────▲───┘
     │                          │
     │  ┌────────┐              │
     └──┤ Y v3.3 ├──────────────┘
        └────────┘
```
* B has version scheme `semver-sepc`
* A has version scheme `pvp`

When resolving dependencies for Y, B v1.3 is evicted in favor of v1.5. This is a valid eviction because B uses `semver-spec`.

For A, while technically A v1.3 appears in the dependency graph, it is never actually depended on in the resolved graph. Indeed, B v1.3 is not there at all anymore. Only B v1.5 was selected, so I expect that only A v1.5 is even *considered*.

**Does the system consider that there was an eviction from A v1.3 to v1.5 in this situation?**

If yes, then it will cause an eviction error (A uses `pvp`), which I *don't* want. In that case, I will have to specify A as using `semver-spec`, even though that would be a lie, in order for the above scenario to be valid. So I'm hoping the answer is No.

*Use case:* A is `scalajs-ir`, B is `sbt-scalajs` or `linker-interface`.

## What about a library that is carefully crafted not to run in binary incompatibilities of its dependency?

I have a lib A, with two running versions v0.6.33 and v1.6.1. A uses semver-spec, so evicting v0.6.33 in favor of v1.6.1 would normally be an eviction error.

However, I also maintain library B, which depends on A v0.6.33. But I carefully crafted B to also be able to link to A v1.x, and do the correct thing.

Therefore, if A v0.6.33 is evicted in favor of v1.x, but the only reason that v0.6.33 was considered is because it was a dependency of B, I *don't* want an eviction error. **How do I signal this to the resolution mechanism?**

*Use case:* A is `sbt-scalajs`, B is `sbt-crossproject-scalajs`.

**Comment by eed3si9n** [comment, marked as answer]
### What if `versionScheme` is different in v1.3 and v1.5 of a lib A?

> If A v1.3 is evicted in favor of A v1.5, what version scheme is used to handle eviction errors? I am hoping that `Some("y")` is used.

A v1.5 is used. According to the implementation, only the winner's attributes are used to extract the versionScheme - https://github.com/sbt/librarymanagement/blob/564119c40407a1663043441059a5f923ad98ab73/core/src/main/scala/sbt/librarymanagement/EvictionError.scala#L87-L94. Thus, so long as the version numbers are increasing, we can change the granularity of compatibility afterwards.

### What about "fake" evictions?

> Does the system consider that there was an eviction from A v1.3 to v1.5 in this situation?

Yes. If both A v1.3 and v1.5 are present in UpdateReport returned by Coursier, it will be considered as an eviction pair.
I can see how this is an area eviction error can improve to be less strict, but it would be difficult to detect the completely irrelevant evictions because we could inject some node in the middle like W v1.0 that depends on the ABI of A v1.3, and it's back in the game.

```
┌────────┐        ┌────────┐
│ A v1.3 │        │ A v1.5 │
└────▲───┘        └───▲────┘
     │                │
     │                │
┌────┴───┐        ┌───┴────┐
│ B v1.3 │        │ B v1.5 │
└────▲───┘        └───▲────┘
     │                │
┌────┴───┐            │    ┌────────┐
│ W v1.0 │            └────┤ X v2.1 │
└────▲───┘                 └────▲───┘
     │                          │
     │  ┌────────┐              │
     └──┤ Y v3.3 ├──────────────┘
        └────────┘
```

What you are uncovering here is that `versionScheme` is more complicated than just the declaration of library author's compatibility intent. Given that the eviction error evaluation takes place at the root of the tree, fully encapsulated implementation JARs should actually _avoid_ declaring its `versionScheme`?

### What about a library that is carefully crafted not to run in binary incompatibilities of its dependency?

> How do I signal this to the resolution mechanism?

I think the traditional way of resolving these types of situation is _not_ including them as a dependency by using `% Provided`. You might have to experiment with it to make sure they do not show up in UpdateReport.

In a special case where carefully crafted library is an sbt plugin it could inject the following as `buildSettings`:

```scala
Seq(libraryDependencySchemes += "org.scala-js" % "sbt-scalajs" % "always")
```

These are all interesting scenarios though!

**Reply by sjrd** [reply]
Thanks @eed3si9n for those detailed answers.

> I can see how this is an area eviction error can improve to be less strict, but it would be difficult to detect the completely irrelevant evictions because we could inject some node in the middle like W v1.0 that depends on the ABI of A v1.3, and it's back in the game.

I would argue that this is still a valid situation. While our ecosystem tooling is such that we can't prevent W from directly referencing anything in A, it would be better to force W to declare a direct dependency on A if it uses stuff from A. That way, in the graph that you post, it would still be a fake eviction with no error. But if W declared a dependency on A, then we would get an actual eviction error, as expected.

> What you are uncovering here is that `versionScheme` is more complicated than just the declaration of library author's compatibility intent. Given that the eviction error evaluation takes place at the root of the tree, fully encapsulated implementation JARs should actually _avoid_ declaring its `versionScheme`?

To be clear, scalajs-ir is definitely not fully encapsulated. There are tools that directly use it, and those would be limited to a minor version. But the regular use case through sbt-scalajs and other stuff depending on sbt-scalajs is not affected by changes in scalajs-ir. I had not thought about *avoiding* declaring a `versionScheme`. It seems that it would be the least bad course of action, given the behavior of artifact resolution. It is definitely better than an *error* for valid situations, and probably better than outright lying by declaring `semver-spec` (which would also hide genuine errors).

> In a special case where carefully crafted library is an sbt plugin it could inject the following as `buildSettings`:
> 
> ```scala
> Seq(libraryDependencySchemes += "org.scala-js" % "sbt-scalajs" % "always")
> ```

IIUC, that would inject the dependency scheme in the build (as in `build.sbt`), which would then apply when resolving library dependencies. But here I want to declare that the sbt plugin's own dependency is fine, so I would need to add something in the *meta-build* (as in `project/plugins.sbt`). Or am I misunderstanding how `buildSettings` works?

**Reply by eed3si9n** [reply]
> But here I want to declare that the sbt plugin's own dependency is fine, so I would need to add something in the *meta-build* (as in `project/plugins.sbt`).

Doh. You're right `buildSettings` won't work.

**Comment by julienrf** [comment]
> If yes, then it will cause an eviction error (A uses `pvp`)

There is one thing that I don’t fully understand in this example: if `A` uses `pvp` and `B` uses `semver-spec`, then when `B` bumps its dependency on `A` from `1.3` to `1.5`, `B` should increment its version number to `2.0`, not `1.5` as in your example?

**Reply by sjrd** [reply]
Normally, if A were a "public" dependency (like cats-effect depends on cats-core) or if A had a high chance of being reused by other stuff (like Guava), that would be desirable, yes. But again, that is only because we do not enforce that directly referencing something in your .class file requires to directly reference that dependency in your POM. If we did enforce that, there would be no reason to require a major version bump in B.

In this case, we have A = scalajs-ir, a library with very niche direct use cases, used as more-or-less internal dependency of B = sbt-scalajs. We can encourage people who directly use that niche API to declare a direct dependency on scalajs-ir. In that case, we won't have any issue and we don't need to bump the major version of sbt-scalajs when we bump the minor version of scalajs-ir.

This is not "fool-proof", of course. But please note that it's a question of trade-offs. If we *did* bump the major version of sbt-scalajs every time an obscure thing in scalajs-ir changes, then we would *actually break many more legit, non-obscure use cases* of other sbt plugins that depend on sbt-scalajs. So we have to choose between:
* being theoretically "fool-proof", and often break legit use cases, and
* being pragmatic, and sometimes break obscure use cases.
